src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -1754,3 +1729,192 @@ Node TheoryStringsRewriter::collectConstantStringAt( std::vector< Node >& vec, u
+              // This is used to partially strip off the endpoint
+              // e.g. str.contains( str.++( "abc", x ), str.++( "cd", y ) ) --> str.contains( str.++( "c", x ), str.++( "cd", y ) )
+              overlap = r==0 ? s.overlap(t) : t.overlap(s);
+            }
+          }else{
+            Assert( ret<s.size() );
+            // can strip off up to the find position
+            // e.g. str.contains( str.++( "abc", x ), str.++( "b", y ) ) --> str.contains( str.++( "bc", x ), str.++( "b", y ) )
+            overlap = s.size()-ret;
+          }
+        }else if( n2[index1].getKind()==kind::STRING_ITOS ){
+          const std::vector<unsigned>& svec = s.getVec();
+          // can remove up to the first occurrence of a non-digit
+          for( unsigned i=0; i<svec.size(); i++ ){
+            unsigned sindex = r==0 ? i : svec.size()-i;
+            if(String::isDigit(svec[sindex])) {
+              break;
+            }else{
+              // e.g. str.contains( str.++( "a", x ), str.to.int(y) ) --> str.contains( x, str.to.int(y) )
+              overlap--;
+            }
+          }
+        }else{
+          //inconclusive
+        }
+        // process the overlap
+        if( overlap<s.size() ){
+          changed = true;
+          if( overlap==0 ){
+            removeComponent = true;
+          }else{
+            // can drop the prefix (resp. suffix) from the first (resp. last) component
+            if( r==0 ){
+              nb.push_back( NodeManager::currentNM()->mkConst(s.prefix(overlap)) );
+              n1[index0] = NodeManager::currentNM()->mkConst(s.suffix(overlap));
+            }else{
+              ne.push_back( NodeManager::currentNM()->mkConst(s.suffix(overlap)) );
+              n1[index0] = NodeManager::currentNM()->mkConst(s.prefix(overlap));
+            }
+          }
+        }
+      }else if( n1[index0].getKind()==kind::STRING_ITOS ){
+        if( n2[index1].isConst() ){
+          CVC4::String t = n2[index1].getConst<String>();
+          const std::vector<unsigned>& tvec = t.getVec();
+          removeComponent = n1.size()>1;
+          // if n1.size()==1, then if n2[index1] contains a non-digit, we can drop the entire component
+          //    e.g. str.contains( str.to.int(x), "123a45") --> false
+          // if n1.size()>1, then if n2[index1] does not contain a digit, we can drop the entire component
+          //    e.g. str.contains( str.++( str.to.int(x), y ), "abc") --> str.contains( y, "abc" )
+          for( unsigned i=0; i<tvec.size(); i++ ){
+            if(String::isDigit(tvec[i])) {
+              if(n1.size()>1) {
+                removeComponent = false;
+                break;
+              }
+            }else{
+              if(n1.size()==1) {
+                removeComponent = true;
+                return true;
+              }
+            }
+          }
+        }
+      }
+      if( removeComponent ){
+        if( n1.size()==1 ){
```

Comment:
Why have the special case for `n1.size() == 1`? It seems like the else branch should work in both cases.

Reply:
The distinct is that the n1.size()==1 case can be more aggressive, since in this case *all* of the string must be in n1[0], and not just a prefix.  Here is an example where we can strip the first component in the n1.size()==1 case but not in the n1.size()>1 case:

str.contains( str.to.int(x), "123a45")  ----> false

but

str.contains( str.++( str.to.int(x), y ), "123a45") -----> str.contains( str.++( str.to.int(x), y ), "123a45") 

(e.g. no change)

The reason is that y may have "a45" as a prefix.

Reply:
Yes, but here we are already in the `removeComponent` case and I think that the else branch does the same thing for `n1.size()==1`:
- `nb/ne.push_back( n1[0] );` is equivalent to `nb/ne.push_back( n1[index0] );` for `n1.size()==1`
- `n1.clear()` is equivalent to `n1.erase( n1.begin(), n1.begin() + 1 );` and `n1.pop_back();` for `n1.size()==1`

Reply:
ohh I see your point (I think I got the code wrong).  The reason I distinguished here is because n1.clear() may be faster than n1.erase( n1.begin(), n1.begin()+1 ), although perhaps this is paranoia.

Reply:
I doubt that there is any difference between the two, especially for vectors of size 1. By the way: `n1.erase( n1.begin(), n1.begin() + 1 );` can be written as just `n1.erase(n1.begin());`.